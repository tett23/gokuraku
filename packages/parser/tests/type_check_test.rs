extern crate parser;

use insta::{assert_debug_snapshot, glob};
use std::fs;

#[test]
fn type_check_test() {
    glob!("../fixtures/pds", "*.pds", |path| {
        dbg!(&path);
        let input = fs::read_to_string(path).unwrap();
        let ast = parser::prose_down_script_parse(&input).unwrap();
        let ir1 = parser::ir::ir1::transform1(&ast);
        let ir2 = parser::ir::ir2::transform2(ir1, ast);

        assert_debug_snapshot!(ir2);
    });
}
