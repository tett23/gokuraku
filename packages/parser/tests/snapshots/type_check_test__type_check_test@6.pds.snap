---
source: packages/parser/tests/type_check_test.rs
expression: ir2
input_file: packages/parser/fixtures/pds/6.pds
---
Ok(
    IR2 {
        ident_symbols: SymbolTable2(
            {
                Ident(
                    "ternary",
                ): (
                    [],
                    Some(
                        TypeAbstructionEnv {
                            trait_constraints: [],
                            handler_constraints: [],
                            abstruction: Arrow(
                                Term(
                                    Constructor(
                                        DataConstructor(
                                            Structual,
                                            ExistsIdent(
                                                ExistsIdent(
                                                    "Int",
                                                ),
                                            ),
                                            [],
                                        ),
                                    ),
                                ),
                                Arrow(
                                    Term(
                                        Constructor(
                                            DataConstructor(
                                                Structual,
                                                ExistsIdent(
                                                    ExistsIdent(
                                                        "Int",
                                                    ),
                                                ),
                                                [],
                                            ),
                                        ),
                                    ),
                                    Arrow(
                                        Term(
                                            Constructor(
                                                DataConstructor(
                                                    Structual,
                                                    ExistsIdent(
                                                        ExistsIdent(
                                                            "Int",
                                                        ),
                                                    ),
                                                    [],
                                                ),
                                            ),
                                        ),
                                        Term(
                                            Constructor(
                                                DataConstructor(
                                                    Structual,
                                                    ExistsIdent(
                                                        ExistsIdent(
                                                            "Int",
                                                        ),
                                                    ),
                                                    [],
                                                ),
                                            ),
                                        ),
                                    ),
                                ),
                            ),
                        },
                    ),
                ),
                Ident(
                    "binary",
                ): (
                    [],
                    Some(
                        TypeAbstructionEnv {
                            trait_constraints: [],
                            handler_constraints: [],
                            abstruction: Arrow(
                                Term(
                                    Constructor(
                                        DataConstructor(
                                            Structual,
                                            ExistsIdent(
                                                ExistsIdent(
                                                    "Int",
                                                ),
                                            ),
                                            [],
                                        ),
                                    ),
                                ),
                                Arrow(
                                    Term(
                                        Constructor(
                                            DataConstructor(
                                                Structual,
                                                ExistsIdent(
                                                    ExistsIdent(
                                                        "Int",
                                                    ),
                                                ),
                                                [],
                                            ),
                                        ),
                                    ),
                                    Term(
                                        Constructor(
                                            DataConstructor(
                                                Structual,
                                                ExistsIdent(
                                                    ExistsIdent(
                                                        "Int",
                                                    ),
                                                ),
                                                [],
                                            ),
                                        ),
                                    ),
                                ),
                            ),
                        },
                    ),
                ),
                Ident(
                    "id",
                ): (
                    [],
                    Some(
                        TypeAbstructionEnv {
                            trait_constraints: [],
                            handler_constraints: [],
                            abstruction: Term(
                                Constructor(
                                    DataConstructor(
                                        Structual,
                                        ExistsIdent(
                                            ExistsIdent(
                                                "Int",
                                            ),
                                        ),
                                        [],
                                    ),
                                ),
                            ),
                        },
                    ),
                ),
                Ident(
                    "unary",
                ): (
                    [],
                    Some(
                        TypeAbstructionEnv {
                            trait_constraints: [],
                            handler_constraints: [],
                            abstruction: Arrow(
                                Term(
                                    Constructor(
                                        DataConstructor(
                                            Structual,
                                            ExistsIdent(
                                                ExistsIdent(
                                                    "Int",
                                                ),
                                            ),
                                            [],
                                        ),
                                    ),
                                ),
                                Term(
                                    Constructor(
                                        DataConstructor(
                                            Structual,
                                            ExistsIdent(
                                                ExistsIdent(
                                                    "Int",
                                                ),
                                            ),
                                            [],
                                        ),
                                    ),
                                ),
                            ),
                        },
                    ),
                ),
            },
        ),
        data_symbols: SymbolTable2(
            {
                ExistsIdent(
                    ExistsIdent(
                        "Bool",
                    ),
                ): Or(
                    Term(
                        Constructor(
                            DataConstructor(
                                Structual,
                                ExistsIdent(
                                    ExistsIdent(
                                        "True",
                                    ),
                                ),
                                [],
                            ),
                        ),
                    ),
                    Term(
                        Constructor(
                            DataConstructor(
                                Structual,
                                ExistsIdent(
                                    ExistsIdent(
                                        "False",
                                    ),
                                ),
                                [],
                            ),
                        ),
                    ),
                ),
                ExistsIdent(
                    ExistsIdent(
                        "Int",
                    ),
                ): Term(
                    Unit,
                ),
                ExistsIdent(
                    ExistsIdent(
                        "True",
                    ),
                ): Term(
                    Unit,
                ),
                ExistsIdent(
                    ExistsIdent(
                        "False",
                    ),
                ): Term(
                    Unit,
                ),
            },
        ),
        handler_symbols: SymbolTable2(
            {},
        ),
        trait_symbols: SymbolTable2(
            {},
        ),
        inst_symbols: SymbolTable2(
            {},
        ),
    },
)
